docs/spec_draft.md

```diff
@@ -1808,6 +1809,68 @@ where `i[d] = j[permutation[d]]`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.while
+
+### Semantics
+
+Produces the output from executing `body` function 0 or more times till the
+`cond` function outputs `true`. More formally, the semantics can be expressed
+using Python-like syntax as follows:
+
+```python
+internal_state = operands
+while cond(internal_state) == True:
+  internal_state = body(internal_state)
+results = internal_state
+```
+
+The behaviour of an infinite loop is undefined.
+
+### Inputs
+
+| Name       | Type                                             |
+|------------|--------------------------------------------------|
+| `operands` | variadic number of tensors of any supported type |
+| `cond`     | `function`                                       |
+| `body`     | `function`                                       |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `results` | variadic number of tensors of any supported type |
+
+### Constraints
+
+  * (C1) `cond` has type `(T0, ..., TN-1) -> tensor<i1>`, where
+         `Ti` = `type(operands[i])`.
+  * (C2) `body` has type `(T0, ..., TN-1) -> (T0, ..., TN-1)`, where
+         `Ti` = `type(operands[i])`.
+  * (C3) For all `i`, `type(results[i])` = `type(operands[i])`.
+
+### Examples
+
+```mlir
+// %constant0: 1
+// %input0: 0
+// %input1: 10
+%results:2 = "stablehlo.while"(%input0, %input1) ({
+^bb0(%argument0: tensor<i32>, %argument1: tensor<i32>):
+  %predicate = "stablehlo.compare"(%argument0, %argument1) {
+    comparison_direction = #stablehlo<comparison_direction LT>
+  } : (tensor<i32>, tensor<i32>) -> tensor<i1>
+  "stablehlo.return"(%predicate) : (tensor<i1>) -> ()
+},  {
+^bb0(%argument0: tensor<i32>, %argument1: tensor<i32>):
+  %output0 = "stablehlo.add"(%argument0, %constant0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
+  "stablehlo.return"(%output0, %argument1) : (tensor<i32>, tensor<i32>) -> ()
+}) : (tensor<i32>, tensor<i32>) -> (tensor<i32>, tensor<i32>)
+// %results#0: 10
+// %results#1: 10
```

Comment:
Proposed reformatting:

```
  %results:2 = "stablehlo.while"(%input0, %input1) ({
    ^bb0(%arg0: tensor<i32>, %arg1: tensor<i32>):
      %0 = "stablehlo.compare"(%arg0, %arg1) {
        comparison_direction = #stablehlo<comparison_direction LT>
      } : (tensor<i32>, tensor<i32>) -> tensor<i1>
      "stablehlo.return"(%0) : (tensor<i1>) -> ()
  }, {
    ^bb0(%arg0: tensor<i32>, %arg1: tensor<i32>):
      %0 = "stablehlo.add"(%arg0, %constant0) : (tensor<i32>, tensor<i32>) -> tensor<i32>
      "stablehlo.return"(%0, %arg1) : (tensor<i32>, tensor<i32>) -> ()
  }) : (tensor<i32>, tensor<i32>) -> (tensor<i32>, tensor<i32>)
```

Feedback:
  1. As far as aligning code goes, let's follow the code example in sort.
  2. `%argN` is an idiomatic way of saying "%argumentN" in MLIR.
  3. I think it's fine to shorten variable names to `%N` when auxiliary ops are involved. Naming inputs and outputs of the primary op is useful because this allows to easily map the spec on them, but everything else - probably not necessary.